nexus/go-handler.md

```diff
@@ -0,0 +1,307 @@
+# Temporal Go SDK - Nexus Handler
+
+## Overview
+
+An addition to the [Temporal Go SDK][temporal-go-sdk] that exposes the ability for users to handle Nexus API requests.
+
+**Goals**
+
+- Expose a simple, high level API for binding Nexus operations to common Temporal constructs
+- Support for starting and canceling operations
+- Support multi-tenant applications with minimal boilerplate
+
+**Non-Goals**
+
+- Support for getting an operation's result or info
+- API definitions for other SDKs
+- API for invoking Nexus calls - refer to the [Nexus Go SDK][nexus-go-sdk] for an out-of-workflow reference client
+- Server implementation details
+- Worker / server API contract
+- Allow advanced users to implement Nexus inbound and outbound request proxies for cross-organization calls where
+  data converters are not shared and data is not exposed to Temporal server.
+
+  This will quickly become a requirement for production use cases, especially with Temporal Cloud's zero trust model. It
+  will be addressed shortly after Nexus becomes available for preview.
+
+> NOTE: These are non-goals for _this_ proposal, but they will be defined in follow up proposals / PRs.
+
+## Background
+
+Nexus is a thin protocol on top of HTTP, defined in [this spec][nexus-http-api]. Temporal serves as an HTTP proxy,
+translating Nexus requests into tasks that can be polled on by an SDK worker on its configured task queue.
+Due to the synchronous nature of HTTP, the generated tasks are not persisted, and require an available worker
+to be processed.
+
+The process of translating requests into tasks and delivering them to workers is out of scope for this proposal.
+
+## SDK Additions
+
+The Temporal Go SDK exposes a `Worker` interface that embeds a `Registry` interface defined as:
+
+```go
+type Registry interface {
+	WorkflowRegistry
+	ActivityRegistry
+}
+```
+
+Add and embed a third registry type: `OperationRegistry` with the following method:
```

Comment:
Embed? So a Nexus handler has to run in a Temporal worker? I wonder if a separate Nexus worker makes sense. I would like to see details on how Nexus handlers are getting work to do.

Reply:
You can run your own Nexus handlers over HTTP but if you want to use Temporal, might as well hook this into a worker.

This is roughly what I have in mind:

![image](https://github.com/temporalio/proposals/assets/52304/9c5221b4-6f70-4ccb-b433-74494fc8869a)

Reply:
While it can be Temporal specific, hooking into an existing worker can be rough in some languages. You may have to do some annoying work to force yourself to be a third poller type in some of the languages as opposed to a separate Nexus worker. I also think the mapping of Temporal task queue to Nexus service may also be a bit rough. But it's ok w/ me if we want to combine these.

Reply:
I think this is better experience for Temporal users.

Reply:
Maybe. I think it's worth showing what a program that only handles operations looks like. Today in the Go SDK for example, an activity worker only has to explicitly disable workflow worker. But other SDKs are more sane here.

Reply:
will users have a way to disable the workflow and activity worker so they can only run a nexus handler?

Reply:
They should be able to IMHO.

Reply:
I'm resolving this. Feel free to unresolve if you think it's worth continuing this thread.